File: web/lib/db/threads.js
// app/lib/db/threads.js — rendered Twitter/X threads
import { pool, safeRows } from "./_core.js";


// ── Twitter/X threads (app/threads.py) — one rendered thread per digest date ──
// Each row is the daily digest rewritten as a ready-to-post thread: a hook tweet
// (carries the /api/og card) + one tweet per bullet, ordered by importance.
export async function listThreadDates() {
  return safeRows(
    `SELECT to_char(digest_date,'YYYY-MM-DD') AS date,
            status,
            jsonb_array_length(tweets) AS tweet_count
     FROM digest_threads
     ORDER BY digest_date DESC`
  );
}


export async function getThread(date) {
  const { rows } = await pool.query(
    `SELECT to_char(digest_date,'YYYY-MM-DD') AS date,
            hook, tweets, cta, og_image_url, model_used, status,
            to_char(created_at,'YYYY-MM-DD"T"HH24:MI:SS') AS created_at
     FROM digest_threads
     WHERE digest_date = $1::date`,
    [date]
  );
  return rows[0] ?? null;
}


export async function latestThreadDate() {
  const { rows } = await pool.query(
    `SELECT to_char(max(digest_date),'YYYY-MM-DD') AS date FROM digest_threads`
  );
  return rows[0]?.date ?? null;
}


// Persist hand-edits from the thread composer. Only the fields supplied are
// written; passing `tweets` replaces the whole array (the editor sends it whole).
export async function updateThread(date, { hook, tweets, cta, status } = {}) {
  const sets = [];
  const vals = [];
  let i = 1;
  if (hook !== undefined)   { sets.push(`hook = $${i++}`);          vals.push(hook); }
  if (tweets !== undefined) { sets.push(`tweets = $${i++}::jsonb`); vals.push(JSON.stringify(tweets)); }
  if (cta !== undefined)    { sets.push(`cta = $${i++}`);           vals.push(cta); }
  if (status !== undefined) { sets.push(`status = $${i++}`);        vals.push(status); }
  if (!sets.length) return null;
  vals.push(date);
  const { rows } = await pool.query(
    `UPDATE digest_threads SET ${sets.join(", ")} WHERE digest_date = $${i}::date
     RETURNING to_char(digest_date,'YYYY-MM-DD') AS date`,
    vals
  );
  return rows[0] ?? null;
}

File: app/db/threads.py
"""app.db.threads — rendered Twitter/X threads (one per digest date)."""
from __future__ import annotations

import json
from datetime import date as date_t

from ._core import _execute, _fetchall, _fetchone


# --------------------------------------------------------------------------- #
# Twitter/X threads (see app/threads.py) — one rendered thread per digest date
# --------------------------------------------------------------------------- #
def upsert_thread(digest_date: "date_t", hook: str, tweets: list[dict], cta: str = "",
                  og_image_url: str = "", model_used: str = "", status: str = "pending") -> None:
    """Upsert the rendered thread for a date. ``status`` defaults to 'pending' (fresh render);
    the pre-publish gate passes 'blocked' when a tweet trips the modality check so the external
    poster (which only picks up 'pending') never ships it."""
    _execute(
        """
        INSERT INTO digest_threads
            (digest_date, hook, tweets, cta, og_image_url, model_used, status, created_at)
        VALUES (%s, %s, %s::jsonb, %s, %s, %s, %s, now())
        ON CONFLICT (digest_date) DO UPDATE SET
            hook         = EXCLUDED.hook,
            tweets       = EXCLUDED.tweets,
            cta          = EXCLUDED.cta,
            og_image_url = EXCLUDED.og_image_url,
            model_used   = EXCLUDED.model_used,
            status       = EXCLUDED.status,
            created_at   = now()
        """,
        (digest_date, hook, json.dumps(tweets), cta, og_image_url, model_used, status),
    )


def get_thread(digest_date: "date_t") -> "dict | None":
    """Return the stored thread for a date (tweets parsed from JSONB) or None."""
    row = _fetchone(
        """SELECT digest_date, hook, tweets, cta, og_image_url, model_used, status, created_at
           FROM digest_threads WHERE digest_date = %s""",
        (digest_date,),
        dict_rows=True,
    )
    return dict(row) if row else None


def get_digest_dates_without_thread() -> list["date_t"]:
    """Digest dates that have bullet analyses but no rendered thread (for backfill)."""
    return [r[0] for r in _fetchall(
        """
        SELECT DISTINCT a.digest_date
        FROM digest_bullet_analysis a
        LEFT JOIN digest_threads t ON t.digest_date = a.digest_date
        WHERE t.digest_date IS NULL
        ORDER BY a.digest_date
        """
    )]
